catalog/to-consul/syncer_test.go

```diff
@@ -246,6 +248,42 @@ func TestConsulSyncer_reapServiceSameNamespace(t *testing.T) {
 	require.Equal("127.0.0.1", service.Address)
 }
 
+// Test that when the syncer is stopped, we don't continue to call the Consul
+// API. This test was added as a regression test after a bug was discovered
+// that after the context was cancelled, we would continue to make API calls
+// to the Consul API in a tight loop.
+func TestConsulSyncer_stopsGracefully(t *testing.T) {
+	t.Parallel()
+
+	// We use a test http server here so we can count the number of calls.
+	callCount := 0
+	consulServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		callCount++
+		// We need to respond with errors to trigger the bug. If we don't
+		// then the code path is only encountered after a timeout which we
+		// won't trigger in the test.
+		w.WriteHeader(500)
+	}))
+	defer consulServer.Close()
+
+	// Start the syncer.
+	client, err := api.NewClient(&api.Config{
+		Address: consulServer.URL,
+	})
+	require.NoError(t, err)
+	s, closer := testConsulSyncer(t, client)
+	s.Sync([]*api.CatalogRegistration{
+		testRegistration("k8s-sync", "bar", "default"),
+	})
+
+	// Compare the call count before and after stopping the server.
+	beforeStopAPICount := callCount
+	closer()
+	time.Sleep(100 * time.Millisecond)
+	// Before the bugfix, the count would be around 500.
+	require.LessOrEqual(t, callCount-beforeStopAPICount, 2)
```

Comment:
Mine was about `1000`; it could be better not to mention it since it probably depends on your machine etc.